### PR TITLE
feat: Track classical register indices for measurements

### PR DESCRIPTION
**Pull request by Altanali**
*Issue #, if available:*
#1004 

*Description of changes:*
Update the BraketProgramContext to extend the changes introduced in https://github.com/amazon-braket/amazon-braket-default-simulator-python/pull/266, which provide a classical register index to the BraketProgramContext `add_measure` method when a measurement assignment is used in an OpenQASM program (e.g. `b[1] = measure q[1];`). This allows translations to and from Braket BDK Circuits and OpenQASM to be consistent with regards to classical register indices used in measurement assignments as currently, these indices are discarded entirely. 

Additionally, updating the `_measure_targets` array in the `Circuit` object was moved to the `add_instruction` method from the `_add_measure` method. This is because users of the Circuit class are _able_ to add measurement instructions via the `add_instruction` method, but `_measure_targets` will not be updated to reflect qubits used in the measurement. This results in incorrect behavior when using `to_ir` to take a circuit with measurements added via `add_instruction` method calls to OpenQASM as `to_ir` will check the `_measure_targets` variable to see if any qubits are measured, see that the list is empty, and apply a measurement to all the qubits in the circuit, including those already measured via the original `add_instruction` method calls. 

*Testing done:*

A round trip test was added to verify that classical register assignments are persisted when going from BDK Circuit to OpenQASM and back to BDK Circuit via the `to_ir` and `from_ir` methods; this is similarly checked when going from OpenQASM to BDK Circuit back to OpenQASM. 

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/unit_tests/braket/circuits/test_circuit.py [hunk 1]
```diff
@@ -891,6 +891,35 @@ def test_from_ir_round_trip_transformation():
...
+                "b[0] = measure q[0];",
+            ]
+        ),
+        inputs={},
+    )
+
+    assert Circuit.from_ir(ir) == Circuit.from_ir(circuit.to_ir("OPENQASM"))
+    assert circuit.to_ir("OPENQASM") == Circuit.from_ir(ir).to_ir("OPENQASM")
+
+
```
Comment (speller26) [inline test/unit_tests/braket/circuits/test_circuit.py:922]: How to get around breaking dependencies:
1. Remove this test from this PR and revert the versions in `setup.py` and `tox.ini`
2. Merge this PR
3. Merge https://github.com/amazon-braket/amazon-braket-default-simulator-python/pull/266
4. Create and merge new PR that reintroduces this test and bumps the default simulator version in `setup.py` to latest release 
```suggestion
```

Comment (Altanali) [reply test/unit_tests/braket/circuits/test_circuit.py]: I've removed the test and reverted the targets for the default simulator version in the setup.py and tox.ini files in the following commit: [7b3ebf3](https://github.com/amazon-braket/amazon-braket-sdk-python/pull/1006/commits/7b3ebf3ffab8e442e0c87d60c003d46cc664b227) !

File: src/braket/circuits/braket_program_context.py [hunk 2]
```diff
@@ -161,16 +162,18 @@ def handle_parameter_value(
...
-    def add_measure(self, target: tuple[int]) -> None:
+    def add_measure(
+        self, target: tuple[int], classical_targets: Iterable[int] | None = None
+    ) -> None:
         """Add a measure instruction to the circuit
 
         Args:
             target (tuple[int]): the target qubits to be measured.
+
+            classical_targets (Iterable[int] | None): the classical registers
```
Comment (speller26) [inline src/braket/circuits/braket_program_context.py:173]:
```suggestion
    def add_measure(
        self, target: tuple[int], classical_targets: Union[Iterable[int], None] = None
    ) -> None:
        """Add a measure instruction to the circuit

        Args:
            target (tuple[int]): the target qubits to be measured.

            classical_targets (Union[Iterable[int], None]): the classical registers
```

File: src/braket/circuits/braket_program_context.py [hunk 3]
```diff
@@ -161,16 +162,18 @@ def handle_parameter_value(
...
-    def add_measure(self, target: tuple[int]) -> None:
+    def add_measure(
+        self, target: tuple[int], classical_targets: Union[Iterable[int], None] = None
+    ) -> None:
         """Add a measure instruction to the circuit
 
         Args:
             target (tuple[int]): the target qubits to be measured.
+
+            classical_targets (Union[Iterable[int], None]): the classical registers
```
Comment (speller26) [inline src/braket/circuits/braket_program_context.py:173]:
```suggestion
    def add_measure(
        self, target: tuple[int], classical_targets: Optional[Iterable[int]] = None
    ) -> None:
        """Add a measure instruction to the circuit

        Args:
            target (tuple[int]): the target qubits to be measured.

            classical_targets (Optional[Iterable[int]]): the classical registers
```

File: src/braket/circuits/braket_program_context.py [hunk 4]
```diff
@@ -161,16 +162,18 @@ def handle_parameter_value(
...
 
-    def add_measure(self, target: tuple[int]) -> None:
+    def add_measure(
+        self, target: tuple[int], classical_targets: Optional[Iterable[int]] = None
+    ) -> None:
         """Add a measure instruction to the circuit
 
         Args:
             target (tuple[int]): the target qubits to be measured.
+
```
Comment (speller26) [inline src/braket/circuits/braket_program_context.py:172]:
```suggestion
```

**Review by speller26 — APPROVED** [review]

**Pull request closed without merging**